wp-includes/sqlite-ast/class-wp-sqlite-information-schema-reconstructor.php

```diff
@@ -0,0 +1,644 @@
+
+		/*
+		 * In MySQL, geometry columns can't have a default value.
+		 *
+		 * Geometry columns are saved as TEXT in SQLite, and in an older version
+		 * of the SQLite driver, TEXT columns were assigned a default value of ''.
+		 */
+		if ( 'geomcollection' === $mysql_type || 'geometrycollection' === $mysql_type ) {
+			return null;
+		}
+
+		/*
+		 * In MySQL, date/time columns can't have a default value of ''.
+		 *
+		 * Date/time columns are saved as TEXT in SQLite, and in an older version
+		 * of the SQLite driver, TEXT columns were assigned a default value of ''.
+		 */
+		if (
+			"''" === $default_value
+			&& in_array( $mysql_type, array( 'datetime', 'date', 'time', 'timestamp', 'year' ), true )
+		) {
+			return null;
+		}
+
+		/**
+		 * Convert SQLite default values to MySQL default values.
+		 *
+		 * See:
+		 *   - https://www.sqlite.org/syntax/column-constraint.html
+		 *   - https://www.sqlite.org/syntax/literal-value.html
+		 *   - https://www.sqlite.org/lang_expr.html#literal_values_constants_
+		 */
+
+		// Quoted string literal. E.g.: 'abc', "abc", `abc`
+		$first_byte = $default_value[0] ?? null;
+		if ( '"' === $first_byte || "'" === $first_byte || '`' === $first_byte ) {
+			return $this->escape_mysql_string_literal( substr( $default_value, 1, -1 ) );
+		}
+
+		// Normalize the default value to for easier comparison.
+		$uppercase_default_value = strtoupper( $default_value );
+
+		// NULL, TRUE, FALSE.
+		if ( 'NULL' === $uppercase_default_value ) {
+			// DEFAULT NULL is the same as no default value.
+			return null;
+		} elseif ( 'TRUE' === $uppercase_default_value ) {
+			return '1';
+		} elseif ( 'FALSE' === $uppercase_default_value ) {
+			return '0';
+		}
+
+		// Date/time values.
+		if ( 'CURRENT_TIMESTAMP' === $uppercase_default_value ) {
+			return 'CURRENT_TIMESTAMP';
+		} elseif ( 'CURRENT_DATE' === $uppercase_default_value ) {
+			return null; // Not supported in MySQL.
+		} elseif ( 'CURRENT_TIME' === $uppercase_default_value ) {
+			return null; // Not supported in MySQL.
+		}
+
+		// SQLite supports underscores in all numeric literals.
+		$no_underscore_default_value = str_replace( '_', '', $default_value );
+
+		// Numeric literals. E.g.: 123, 1.23, -1.23, 1e3, 1.2e-3
+		if ( is_numeric( $no_underscore_default_value ) ) {
+			return $no_underscore_default_value;
+		}
+
+		// HEX literals (numeric). E.g.: 0x1a2f, 0X1A2F
+		$value = filter_var( $no_underscore_default_value, FILTER_VALIDATE_INT, FILTER_FLAG_ALLOW_HEX );
+		if ( false !== $value ) {
+			return $value;
+		}
+
+		// BLOB literals (string). E.g.: x'1a2f', X'1A2F'
+		// Checking the prefix is enough as SQLite doesn't allow malformed values.
+		if ( str_starts_with( $uppercase_default_value, "X'" ) ) {
+			// Convert the hex string to ASCII bytes.
+			return "'" . pack( 'H*', substr( $default_value, 2, -1 ) ) . "'";
+		}
+
+		// Unquoted string literal. E.g.: abc
+		return $this->escape_mysql_string_literal( $default_value );
```

Comment:
Could this happen in practice? Or should this be an error? I'm worried about converting things like `DEFAULT NOW()` to `DEFAULT "NOW()"`

Reply:
@adamziel Unfortunately, `DEFAULT abc` is valid:

```sql
CREATE TABLE t (id INT, value TEXT DEFAULT abc);
INSERT INTO t (id) VALUES (1);
SELECT * FROM t;
-- id: 1, value: 'abc'
```
And unfortunately, `PRAGMA table_info(t)` doesn't normalize it and returns it without the quotes.

That said, `DEFAULT NOW()` is a syntax error, so I suppose the unquoted syntax is only for whatever would be a valid unquoted identifier. SQLite only supports the handled date/time functions at the moment: https://www.sqlite.org/syntax/literal-value.html

I could check for sth like if the string doesn't have quotes or other special chars, but maybe it's not worth it.

Reply:
Oh wow, TIL! Good find @JanJakes 🤯 What are the rules for a literal? e.g. I suppose it cannot contain opening parenthesis `(`. Perhaps we can add a few safeguards here to only let through the values that would parse correctly?

Reply:
The only reason I can think of to make it worth it is to throw an explicit error when it happens. We'd just be kind to someone debugging a potential issue with this in the future. It doesn't seem very likely to happen, though, so maybe we could skip it here if it would require a lot of research.

Reply:
I'll resolve this conversation, I think we're good as far as this PR goes. Thank you.